File: 2021/02/07/beomjo/algorithm/BOJ/java/17298.java
import java.io.*;
import java.util.Stack;
import java.util.StringTokenizer;

class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        int N = Integer.parseInt(br.readLine());
        int[] answer = new int[N];
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        Stack<NGE> s = new Stack<>();

        for (int i = 0; i < N; i++) {
            int number = Integer.parseInt(st.nextToken());
            while (!s.isEmpty() && number > s.peek().element) {
                answer[s.pop().position] = number;
            }
            s.push(new NGE(number, i));
        }

        for (int r : answer) {
            if (r == 0) r = -1;
            bw.write(r+" ");
        }
        bw.flush();
    }
}

class NGE {
    int element, position;

    public NGE(int element, int position) {
        this.element = element;
        this.position = position;
    }
}

File: 2021/02/07/beomjo/algorithm/BOJ/kotlin/17298.kt
import java.util.*

fun main() {
    val N = readLine()!!.toInt()
    val input = readLine()!!.split(" ").map { it.toInt() }
    val stack = Stack<NGE>()
    val answer = MutableList(N) { -1 }

    repeat(N) { i ->
        val number = input[i]
        while (!stack.isEmpty() && number > stack.peek().element) {
            answer[stack.pop().position] = number
        }
        stack.push(NGE(number, i))
    }

    print(answer.joinToString(" "))
}

data class NGE(
    val element: Int,
    val position: Int,
)